### PR TITLE
feat: add retry to agol publish calls

### LINKED ISSUE #109: Add retry to agol.py
To hopefully help with these types of errors:
image

### PR DESCRIPTION
**Pull request by stdavis**
Closes #109

Here's what a test of the retry logic looked like:
image

We could add it to other methods as well if they become a problem.

### CODE REVIEW

**Comment by stdavis** [comment]
@jacobdadams The way that I think of lambdas in python is anonymous functions. Thanks for the review.

File: sgid/agol.py [hunk 1]
```diff
@@ -19,6 +20,21 @@
 DRAFTS_FOLDER_NAME = 'sddrafts'
 
 
+def retry(worker, logger, tries=1):
+    max_tries = 5
```
Comment (steveoh) [inline sgid/agol.py:24]: **thought**: 5 tries for a 1 minute of sleeping seems like a lot of time to give a service to respond correctly. I'm ok with it but wonder if one fails all will fail and the time might stack up.

File: sgid/agol.py [hunk 2]
```diff
@@ -86,7 +102,7 @@ def publish(self, title, feature_class):
         source_item.protect()
 
         self.log.debug('publishing feature service')
-        item = source_item.publish()
+        item = retry(lambda: source_item.publish(), self.log)
```
Comment (steveoh) [inline sgid/agol.py:105]: **question**: do you need lambda? 
```suggestion
        item = retry(source_item.publish, self.log)
```

**Comment by steveoh** [comment]
> I've never seen that way of creating a generic interface for passing any function by wrapping it in a lambda function (if I'm understanding the code correctly). Cool!

It's a very handy technique. A lot of people don't realize that you can pass functions as arguments in python. They can really help to remove code duplication.

Scott is using it in a funny way here though. Since python can pass a function by reference and the function in question doesn't have any parameters, the function can be passed without the `()` and executed in the other method. If you were to expand the lambda as it is written it would be like

```py
def anon_function():
    source_item.publish()
```

which is basically the same as `source_item.publish()`. Hopefully that highlights the redundancy.

Lambda's are mainly used for one liners e.g. `function_reference = lambda x: x + 1`. These are mostly appropriate for simple things like sorting or filtering functions. Inline function are typically more refactor friendly since they don't have any limitations. For instance lambda's can not have multiple expressions. So if you wanted to add more logic to the `lambda: source_item.publish()` you would have to convert it to an inline method.

**Comment by jacobdadams** [comment]
Yeah, passing functions as args was new to me too. I've used lambdas once or twice for custom sorts. I was wondering, in this instance, if passing the lambda allows extensibility because now that whatever function you're passing is encased in a lambda it can have any number of arguments without altering `retry()`? Or does the signature of the function you're passing in as a parameter get "bundled up" (for lack of a better term) and remembered for when you call `function_argument()` later in the higher-order function?

File: sgid/agol.py (hunk 1 above)
Comment (stdavis) [reply sgid/agol.py:24]: That's a good point. I cut it down to three tries.

**Comment by stdavis** [comment]
> are you thinking we might move this to the seat module of forklift or something?

We could if we feel like it would be useful for other pallets. It might be good to be in its own package since it's not something specific to forklift. Although at that point it might be just as easy to use an existing library.

**Comment by stdavis** [comment]
> I was wondering, in this instance, if passing the lambda allows extensibility because now that whatever function you're passing is encased in a lambda it can have any number of arguments without altering `retry()`? Or does the signature of the function you're passing in as a parameter get "bundled up" (for lack of a better term) and remembered for when you call `function_argument()` later in the higher-order function?

I would say that my `retry` in this case is expecting a function with zero parameters. That's why I'm wrapping some in lambdas. That way they don't require any parameters.

**Pull request merged**